### PR TITLE
[Examples/Front-end] Tic-tac-toe

### PR DESCRIPTION
**Pull request by amnn**
## Description

Build front-ends for shared and multi-sig tic-tac-toe, using
`create-react-dapp`.

## Test plan

Manually tested (screenshots below).

#### New Shared game
shared game

#### New MultiSig game
multi-sig game

#### Game in-progress
game in-progress

#### Game won, viewed by player
game win

#### Game won, viewed by spectator (no wallet connected)
game win spectator

#### Delete game dialog
delete game

## Stack

- #18525

### CODE REVIEW

**Comment by amnn** [comment]
@hayes-mysten, as promised, here's the experience report from modernising this demo!

I'm going to focus on the areas where I struggled, but I wanted to start by saying that dApp Kit, Radix UI, React, and React Query were great -- they significantly sped things up for me, and I'm confident I wouldn't have been able to deliver anything near the quality I was able to without them. A big 👍 for starting with an opinionated stack of tools and libraries that work well together, because they taught me a lot about what's available, and what the best practices are.

Some of the things I found tricky are specific to me being a front-end/TypeScript noob. I tried to filter those things out, but I may not have succeeded (don't know what I don't know!), but here goes:

#### Wallet <-> NetworkConfig syncing
This was an early stumbling block because NetworkConfig defaults to testnet and I was building on localnet, so without a way to switch networks, I was a bit stuck. Initially, I was able to unblock myself by changing the default to localnet, but this may not be a sensible default. Eventually, I added some logic to query the current account, its network, check whether we have a config for that network, and set it if so (or error otherwise).

I wonder whether we can package this behaviour up as a default? So that out of the box, the app will switch its config according to the network the wallet sets.

Similarly, I had to roll my own script to deploy a package, and update the dApp's configs to pick up the newly deployed package -- is this something we can bake into `create-react-dapp`? (I used a bash script, but I imagine if we were to do it in `create-react-dapp`, it would be written in TypeScript).

#### Address <-> App Stickiness
This one's a known issue, I got stuck trying to connect some other address to the app during testing. My understanding is that you need to go into the wallet settings and dissociate the app from that address and then reconnect.

My expectation here is that once you connect a wallet, you should be able to select addresses from that wallet to connect to the app using the drop-down, but only one address ever shows here.

#### Query Invalidation
Main issue here was that it was not clear what the default query key was -- I was supplying a key as a parameter, but I didn't realise that it was being prefixed with some other stuff. There's a couple of ways I could see this being solved:

- Docs -- a doc comment would have saved a lot of trouble here.
- Not adding the prefix if the queryKey parameter is set.
- Abstracting away query invalidation (this is what I ended up doing). I ended up adopting a pattern where query functions returned a tuple -- one element being the results, and the other being a function that invalidates the query. I found this pretty easy to work with, but it maybe falls down if you want to invalidate multiple queries simultaneously. Maybe a middle ground is to have the function return the query key it used?

#### Docs in LSP
Finding docs in the LSP was a struggle, and it is made more complicated by the fact that the symbols the language server like to jump to are usually just type declarations, and there are some pretty exotic types here. This wasn't a problem that was unique to our libraries though -- while the standard library and maybe core React hooks had good documentation, many other libraries didn't, which is a shame.

#### Transforming an Object on Load
When querying the Game object over RPC, I needed to perform some extra checks and transformations. In some cases the checks would yield errors. I ended up doing this in a custom hook, but it might be nice to allow users to provide a transformation function in a call to `useSuiClientQuery`, (it's particularly useful when the extra work/transformation may yield an error).

#### Radix UI is now at 3.x
`create-react-dapp` depends on Radix at `2.x`, but it's at `3.x` now. Radix doesn't keep around legacy docs so I ended up upgrading it to make sure the APIs matched what I was seeing on their site -- can we bump it in the template?

#### Importing modules
I found the pattern of always using relative imports to be quite jarring -- is that the best practice? In the end, I installed a vite plugin that would let me depend on modules relative to the `src` directory, which I found much more manageable when dealing with dependencies across different sub-directories. If that seems reasonable, can we add that template and the associated `baseUrl` configuration to the template?

#### Aliased types
There are some types (the ones I can think of are `MultiSigPlublicKey` and `PublicKey`) that appear under `sui/client`, `sui/cryptography`, and `sui/multisig`. This means that sometimes the language server can end up suggesting the wrong one for import -- it would be great if we could deduplicate these.

#### `vector<u8>` transaction inputs
The new `tx.pure` API is miles better than what it was before, but I still ended up tripping over the old issue of whether or not to pre-serialize an input when I was holding a `Uint8Array` and I wanted to pass it to a `vector<u8>` parameter. In the end, I manually serialized it using the `bcs` library, but it seems reasonable to add a `tx.pure.bytes` or something for this case?

File: pnpm-workspace.yaml
Comment (manolisliolios) [reply pnpm-workspace.yaml:1]: I usually add the `--ignore-workspace` flag  (`pnpm install --ignore-workspace`) for non-workspace projects in the monorepo!

File: examples/tic-tac-toe/ui/src/hooks/useTransactions.ts
Comment (manolisliolios) [reply examples/tic-tac-toe/ui/src/hooks/useTransactions.ts:48]: I think what you did is the recommended way. @hayes-mysten to confirm!

File: examples/tic-tac-toe/ui/src/hooks/useTransactions.ts
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/hooks/useTransactions.ts:72]: It's a good react pattern to use `mutations` for any "POST" style transactions (user actions)! 

That helps:
1. Keep all mutations in one place
2. Add easier loading states / error handling.

Check this example for some mutations: https://github.com/MystenLabs/sui/blob/d5e1b5fa8625e1f75df511e62453d25b862e94e0/examples/trading/frontend/src/mutations/escrow.ts#L15

Comment (amnn) [reply examples/tic-tac-toe/ui/src/hooks/useTransactions.ts:72]: This particular hook doesn't perform any mutations, it just builds transactions, but it sounds like my `useExecutor` hook is a mutation (it is built on top of `useSignAndExecuteTransaction` which is itself built on `useMutation`, I believe).

File: pnpm-workspace.yaml
Comment (amnn) [reply pnpm-workspace.yaml:1]: Nice, that worked, thanks for the pointer!

File: examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx
Comment (amnn) [inline examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx:27]: Here `useCurrentAccount` has returned some bytes for the public key, and they seem to be raw -- how would a wallet return the public key when the signature scheme was not ED25519?

File: examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx:111]: I can't believe it's my turn on this comment. Not rust here... 😆 
```suggestion
	if (address === '') {
```

File: examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx:98]: super nit: can skip the "else" statement and just return <Box />.

Theoretically they can all be top-level ifs without `if,elseif,else` but that might be my personal preference.

File: examples/tic-tac-toe/ui/src/config.ts
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/config.ts:11]: Looking at this, should we create a super simple parser for these configs looking for `.env.{network}.json` which will automatically read these? So if someone is calling has first called `./publish.sh`, they won't need to come here and import themselves.

File: examples/tic-tac-toe/ui/src/hooks/useTrophyQuery.ts
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/hooks/useTrophyQuery.ts:39]: What might be convenient is to export all `queryKey` names so you could theoretically invalidate keys more generic throughout the app, if you needed to. E.g. `QueryKeys { GameEndState, ...}`, instead of having to expose an "invalidate" function on each of these.

Returning the "useQuery" hook as is, also exports a `refetch` function that you can use to re-fetch the values, instead of re-defining a new `invalidate` function on every hook.

File: examples/tic-tac-toe/ui/src/App.tsx
Comment (manolisliolios) [reply examples/tic-tac-toe/ui/src/App.tsx:24]: To the depth of my knowledge, that's the way!

File: examples/tic-tac-toe/ui/src/main.tsx
Comment (manolisliolios) [inline examples/tic-tac-toe/ui/src/main.tsx:26]:
```suggestion
						<>
```
nit: Skips 1 level of nesting

**Review by manolisliolios — APPROVED** [review]
Thanks a lot Ashok. Approving to unblock, overall this looks amazing. UX is super smooth! 🚀 

Added some comments / suggestions here and there!

File: examples/tic-tac-toe/ui/src/env.localnet.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/env.localnet.ts:1]: Yep, maybe I'll add some dummy values here and then add a .gitignore for localnet.

File: examples/tic-tac-toe/ui/src/config.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/config.ts:11]: I can stub these out as files, instead -- that might be the easiest way.

Comment (amnn) [reply examples/tic-tac-toe/ui/src/config.ts:12]: I got this to mostly work -- I stubbed out `env.localnet.ts` and locally I'm using `--skip-worktree` locally to avoid updating it -- not great, but it'll do for now.

File: examples/tic-tac-toe/ui/src/hooks/useGameQuery.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/hooks/useGameQuery.ts:49]: I like option (1) a fair bit, but it doesn't play too nicely with how `dapp-kit` currently handles keys (you can only supply a key suffix, not a prefix) -- I wonder if there's a way we can tweak the way `dapp-kit` works to allow these kind of patterns, cc @hayes-mysten . (2) seems less ideal to me, because it sounds like we would switch things around to make refetching the default, and then have to manually turn it off in places, which would work for our case here, but I'm not sure how generalisable that is.

What I was originally hoping was that there was a way to describe the data dependency, so that `react-query` could detect that there was a change in the query input and therefore re-run it, in which case we could treat our game as a redundant input to all the other queries, and use that mechanism to trigger the refetches -- is that feasible at all?

File: examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx
Comment (amnn) [reply examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx:111]: touché 😀 `trim` is guaranteed to return a `string` though, right?

File: examples/tic-tac-toe/ui/src/hooks/useTrophyQuery.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/hooks/useTrophyQuery.ts:39]: I thought about this, but I think again it kind of runs counter to the way `dapp-kit` is working, which abstracts away the query keys in its queries. Maybe if we can align on that, this would be a good way to go.

**Comment by amnn** [comment]
Just finishing up addressing the other comments, but I'm noticing that something is still not quite right with the workspace set-up, so could use some advice on that @manolisliolios / @hayes-mysten .

It looks like `pnpm install --frozen-lockfile` still believes the `package.json` in `examples/tic-tac-toe/ui` is part of the workspace, even though it has its own lock-file.

I was trying to follow the path that the `trading` example took, which didn't seem to include anything to exclude it from the workspace, but now I'm wondering whether that was just a happy coincidence -- i.e. it happens to use all the same dependencies.

Is the right thing to do here to exclude all the examples from the workspace, and ensure they each have their own lock files?

File: examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx
Comment (amnn) [reply examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx:98]: The `<Box />` is necessary to make sure there's something on the left hand side of the flex box -- otherwise the "Play" button this is sitting next to will get pulled to the left when there are no validation issues -- but I'll switch this to an early return style!

Comment (manolisliolios) [reply examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx]: Oh that's exactly what I meant (early return versus nested if-elseif-else statements). But that's just me not liking nesting 😄

Comment (amnn) [reply examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx]: Ah, got it!

Comment (amnn) [reply examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx]: I thought you were saying just `return;` (i.e. don't need to return anything instead of `Box`), but I understand now!

File: examples/tic-tac-toe/ui/src/hooks/useTransactions.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/hooks/useTransactions.ts:72]: Made these changes `useExecutor` is moved to the `mutations` directory, and it exposes the various status-related fields. It felt proper to make use of these, so I added some logic to handle pending transactions for "Play" buttons.

File: examples/tic-tac-toe/ui/src/hooks/useTransactions.ts
Comment (hayes-mysten) [reply examples/tic-tac-toe/ui/src/hooks/useTransactions.ts:48]: This is mostly the correct way to do this, but the last toBytes isn't needed: `tx.pure(bcs.vector(bcs.u8()).serialize(admin.toRawBytes()))`



`bcs.bytes()` is an existing method that is used for fixedLength bytes, so I think also calling vector<u8> bytes would be confusing.

I just opened a PR to make this a little better though: https://github.com/MystenLabs/sui/pull/18561/files

This will add `tx.pure.vector('u8', admin.toRawBytes())`

File: examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx
Comment (hayes-mysten) [reply examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx:27]: the walletStandard doesn't expose enough information to make this work.  Generally dapps are built around addresses, not public keys, and some wallet accounts won't return meaningful publicKey bytes (eg stashed).

That being said, you theoretically could try each supported schema, and see if it produces a pubKey that has the same address as the account

File: examples/tic-tac-toe/ui/src/hooks/useGameQuery.ts
Comment (hayes-mysten) [reply examples/tic-tac-toe/ui/src/hooks/useGameQuery.ts:49]: Going to defer to @williamrobertson13 here who has some more react/front-end expertise

Comment (hayes-mysten) [reply examples/tic-tac-toe/ui/src/hooks/useGameQuery.ts:49]: this is definitely something we should have a good (and well documented) solution for

File: examples/tic-tac-toe/ui/src/hooks/useTurnCapQuery.ts
Comment (hayes-mysten) [inline examples/tic-tac-toe/ui/src/hooks/useTurnCapQuery.ts:41]: never seen this used instead of just doing a while, had no idea this would work, `undefined`, With the condition being empty, I would have expected this to never loop..  TIL

**Comment by hayes-mysten** [comment]
> Wallet <-> NetworkConfig syncing
This is definitely an issue.  I talked with @Jordan-Mysten about this yesterday.
We will be updating dapp-kit to properly send the chain identifier so that the dapps selected network gets used when building/signing/executing transactions
* The extension wallet doesn't currently respect this option, so this won't work until we re-unify the code-bases
* Dapp-kit doesn't currently know about chain-identifiers so we will need to update it to have users configure chain identifiers as part of the network config

We don't want to sync settings between the wallet and dapp, but we want the dapp state to be authoratitive, so that the wallet state doesn't matter.

> Similarly, I had to roll my own script to deploy a package, and update the dApp's configs to pick up the newly deployed package -- is this something we can bake into create-react-dapp? (I used a bash script, but I imagine if we were to do it in create-react-dapp, it would be written in TypeScript).

We'd like to do something around this.  I think the current plan was to wait for the package registry, and build a good developer workflow on top of that.  Eventually this should be part of the dapp-starter, but we have't built it yet.  Adding in a simple bash script as a temporary solution might also be good. 

> Address <-> App Stickiness
This one's a known issue, I got stuck trying to connect some other address to the app during testing. My understanding is that you need to go into the wallet settings and dissociate the app from that address and then reconnect.

> My expectation here is that once you connect a wallet, you should be able to select addresses from that wallet to connect to the app using the drop-down, but only one address ever shows here.

The dapp can only access the addresses that you selected during the connection flow.  This isn't intuitive, but we will be changing the default to select all addresses, so that you won't have to disconnect/reconnect to see other wallet addresses in the dapp.   The mobile wallet (which will become the extension wallet soon) already clears the connection state when you disconnect from the dapp side.  The extension wallet preserving the connection state when the dapp disconnects is part of what makes this confusing today.

> Query Invalidation
Main issue here was that it was not clear what the default query key was -- I was supplying a key as a parameter, but I didn't realise that it was being prefixed with some other stuff. There's a couple of ways I could see this being solved:

We should definitely address this. I'm going to bring this up with people who know more about react than I do

> Docs in LSP
Finding docs in the LSP was a struggle, and it is made more complicated by the fact that the symbols the language server like to jump to are usually just type declarations, and there are some pretty exotic types here. This wasn't a problem that was unique to our libraries though -- while the standard library and maybe core React hooks had good documentation, many other libraries didn't, which is a shame.
This is something we can probably improve.  Jumping to type-definitions instaed of sources is probably not something we can resolve, but we can introduce more typedoc comments throughout the codebase

> Transforming an Object on Load
When querying the Game object over RPC, I needed to perform some extra checks and transformations. In some cases the checks would yield errors. I ended up doing this in a custom hook, but it might be nice to allow users to provide a transformation function in a call to useSuiClientQuery, (it's particularly useful when the extra work/transformation may yield an error).

You can pass a `select` callback in the options of useSuiClientQuery that will take/transform the data (as long as what you are doing is synchronous)

> Radix UI is now at 3.x
create-react-dapp depends on Radix at 2.x, but it's at 3.x now. Radix doesn't keep around legacy docs so I ended up upgrading it to make sure the APIs matched what I was seeing on their site -- can we bump it in the template?

Just bumped the version, thanks for calling this out

> Importing modules
I found the pattern of always using relative imports to be quite jarring -- is that the best practice? In the end, I installed a vite plugin that would let me depend on modules relative to the src directory, which I found much more manageable when dealing with dependencies across different sub-directories. If that seems reasonable, can we add that template and the associated baseUrl configuration to the template?

using some sort of root alias has become a lot more common recently, but relative imports have been the standard in the JS ecosystem for a long time. 
I'd be happy to add something like this to the starter templates though

> Aliased types
There are some types (the ones I can think of are MultiSigPublicKey and PublicKey, but there may be more) that appear under sui/client, sui/cryptography, and sui/multisig. This means that sometimes the language server can end up suggesting the wrong one for import -- it would be great if we could deduplicate these.

This is definitely a pain.  I've removed so many already but `@mysten/sui/client` includes a lot of generated types from JSON RPC.  This will oneday go away when we are off of JSON RPC (or maybe we can just break it out into a separate pacakge) but removing all the duplicates is pretty tough.   We might be able to prefix the generated types at some point in the future, but it would require another round of breaking changes that the community probably wouldn't be happy with

> vector<u8> transaction inputs
The new tx.pure API is miles better than what it was before, but I still ended up tripping over the old issue of whether or not to pre-serialize an input when I was holding a Uint8Array and I wanted to pass it to a vector<u8> parameter. In the end, I manually serialized it using the bcs library, but it seems reasonable to add a tx.pure.bytes or something for this case?

Left a comment about this in one of the other threads on this PR

> Wallet Standards's signAndExecuteTransaction vs SuiClient.signAndExecuteTransaction
I didn't actually trip up on this one during my own development, but noticed someone on [Telegram](https://t.me/c/1742460518/25468) had this issue recently and got tripped up on it that way.

> When would someone choose to use the execution via wallet standard vs their own SuiClient? It seems like we're offering two similar strategies for wallets and apps to communicate with each other, but I can't really see a clear differentiator (where one is more appropriate than the other), and the choice itself can be confusing.

The TS sdk isn't just used for front end apps connected to wallets.  suiClient.signAndExecuteTransaction is for signing with a keypair, vs `useSignAndExecuteTransaction` which is used to sign via the wallet.  

> Wallet Standard Public Keys
The Wallet Standard returns public keys as raw bytes, and as far as I could tell, there was not a way to figure out what the signature scheme was for that public key. In this demo, I assumed ED25519, but if a wallet is using some other kind of key, this could spell trouble. We have a byte form that is annotate ("Sui" Bytes), can we use that instead?

The public keys in the wallet standard aren't very useful and are better to ignore.  I commented on this somewhere else, but we can't meaningfully provide them for all wallets, and as you noticed there isn't enough information to derive a publicKey instance from the bytes. They are part of the wallet-standard that we don't control.  Generally we should just use the address and ignore the pubkey

File: examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx
Comment (amnn) [reply examples/tic-tac-toe/ui/src/components/NewMultiSigGame.tsx:27]: Would it be feasible to have the wallet standard return a public key in "SuiBytes" form? (I believe it's just an enum, so it's the raw bytes prefixed with a variant index), or add an optional signature scheme field, so I could write something like:

```
const { address, publicKey: bytes, keySignature } = useCurrentAccount() || {};
const publicKey = keySignature && bytes && publicKeyFromRawBytes(keySignature, bytes);
```

File: examples/tic-tac-toe/ui/src/hooks/useTurnCapQuery.ts
Comment (amnn) [reply examples/tic-tac-toe/ui/src/hooks/useTurnCapQuery.ts:41]: I believe this is a C idiom and it exists for the sole reason that it's the shortest string of characters to achieve this result:

```
while (true) {
while (1) {
for (;;) {
```

But I'm not attached to it -- if it looks weird, I'll switch it to whatever's the convention!

**Comment by amnn** [comment]
Amazing, thanks for taking a look, and for the thorough response @hayes-mysten !

**Comment by amnn** [comment]
In the end, while trying to extricate the tic-tac-toe UI from the workspace, I realised that all the other examples are explicitly in the workspace, so it's a bit out of scope of this PR to tackle that issue. If it's desirable to move examples out of the workspace, we should come back to tackle that.

**Comment by amnn** [comment]
...that went horribly wrong, and I take that all back. Adding tic-tac-toe to the workspace broke everything, so in the end it was much easier to exclude it explicitly, let's hope that sticks.

**Pull request merged**
